core/src/callback.jl

```diff
@@ -393,22 +398,131 @@ function set_fractional_flow_in_allocation!(
-        if is_active(p.allocation) && node isa FractionalFlow && field == :fraction
-            set_fractional_flow_in_allocation!(p, node_id, value)
-        end
+function update_parameters!(
+    tabulated_rating_curve::TabulatedRatingCurve,
+    parameter_update::ParameterUpdate,
+)::Nothing
+    if parameter_update.node_type != NodeType.TabulatedRatingCurve
+        return nothing
+    end
+    (; node_idx, active, table) = parameter_update
+    tabulated_rating_curve.active[node_idx] = active
+    if !isempty(table.t)
+        tabulated_rating_curve.tables[node_idx] = table
+    end
+    return nothing
+end
+
+function update_parameters!(
+    fractional_flow::FractionalFlow,
+    parameter_update::ParameterUpdate,
+    p::Parameters,
+)::Nothing
+    if parameter_update.node_type != NodeType.FractionalFlow
+        return nothing
     end
+    (; node_idx, fraction) = parameter_update
+    fractional_flow.fraction[node_idx] = fraction
+
+    if is_active(p.allocation)
+        set_fractional_flow_in_allocation!(p, fractional_flow.node_id[node_idx], fraction)
+    end
+    return nothing
+end
+
+function update_parameters!(
+    pid_control::PidControl,
+    parameter_update::ParameterUpdate,
+)::Nothing
+    if parameter_update.node_type != NodeType.PidControl
+        return nothing
+    end
+    (; node_idx, active, target, pid_params) = parameter_update
+    pid_control.active[node_idx] = active
+    if !isempty(target.t)
+        pid_control.target[node_idx] = target
+    end
+    if !isempty(pid_params.t)
+        pid_control.pid_params[node_idx] = pid_params
+    end
+    return nothing
+end
+
+function update_parameters!(
+    linear_resistance::LinearResistance,
+    parameter_update::ParameterUpdate,
+)::Nothing
+    if parameter_update.node_type != NodeType.LinearResistance
+        return nothing
+    end
+    (; node_idx, active, resistance) = parameter_update
+    linear_resistance.active[node_idx] = active
+    if !isnan(resistance)
+        linear_resistance.resistance[node_idx] = resistance
+    end
+    return nothing
+end
+
+function update_parameters!(
+    manning_resistance::ManningResistance,
+    parameter_update::ParameterUpdate,
+)::Nothing
+    if parameter_update.node_type != NodeType.ManningResistance
+        return nothing
+    end
+    (; node_idx, active, manning_n) = parameter_update
+    manning_resistance.active[node_idx] = active
+    if !isnan(manning_n)
+        manning_resistance.manning_n[node_idx] = manning_n
+    end
+    return nothing
+end
+
+function set_control_params!(p::Parameters, node_id::NodeID, control_state::String)::Nothing
+    parameter_update = p.discrete_control.control_mapping[(node_id, control_state)]
+
+    # Only the method corresponding to the type of the node_id actually updates parameters,
+    # but this way there is no runtime dispatch on node type
+    update_parameters!(p.pump, parameter_update)
+    update_parameters!(p.outlet, parameter_update)
```

Comment:
Consider renaming either the function/argument/struct to indicate the parameter update happen because of control.